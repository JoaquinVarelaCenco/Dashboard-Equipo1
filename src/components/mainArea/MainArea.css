/* code css */
.mainArea{
    background-color: #242424;
    width: 100%;
}

.clear{
    background-color: #b1b1b1;
}

.dark{
    background-color: #242424;
}

.mainArea__content{
    margin-top: 72px;
    min-height: calc(100vh - 72px);
}