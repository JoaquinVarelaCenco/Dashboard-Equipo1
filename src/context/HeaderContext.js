import { useState, createContext } from "react";

export const HeaderContext = createContext("home");

export const HeaderProvider = ({ children }) => {
  const [page, setPage] = useState("home");

  const value = {
    page,
    currentPage: (p) => setPage(p),
  };

  return (
    <HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
  );
};
