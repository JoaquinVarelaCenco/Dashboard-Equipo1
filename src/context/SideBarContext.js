import { useState, createContext } from 'react'

export const SideBarContext = createContext('unhide');

export const SideBarProvider = ({children}) => {

    const [ visibility, setVisibility ] = useState('unhide');

    const value = {
        visibility,
        toggleVisibility: () => setVisibility(prev => prev === 'unhide' ? 'hide' : 'unhide')
    }


return (
        <SideBarContext.Provider value={value}>
            {children}
        </SideBarContext.Provider>
    )
}