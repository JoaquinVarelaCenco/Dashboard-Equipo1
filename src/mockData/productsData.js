module.exports = [
    {
      "id": 2,
      "title": "iPhone X",
      "description": "SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip with ...",
      "price": 899,
      "rating": { "rate": 4.44, "count": 126 },
      "stock": 34,
      "category": "smartphones",
      "images": [],
      "lastModified": "2022-10-12T01:07:11.630Z"
    },
    {
      "id": 3,
      "title": "Samsung Universe 9",
      "description": "Samsung's new variant which goes beyond Galaxy to the Universe",
      "price": 1249,
      "rating": { "rate": 4.09, "count": 845 },
      "stock": 36,
      "category": "smartphones"
    },
    {
      "id": 23,
      "title": "Orange Essence Food Flavou",
      "description": "Specifications of Orange Essence Food Flavour For Cakes and Baking Food Item",
      "price": 14,
      "rating": { "rate": 4.85, "count": 1645 },
      "stock": 26,
      "category": "groceries",
      "images": [
        "https://dummyjson.com/image/i/products/23/1.jpg",
        "https://dummyjson.com/image/i/products/23/2.jpg",
        "https://dummyjson.com/image/i/products/23/3.jpg",
        "https://dummyjson.com/image/i/products/23/4.jpg",
        "https://dummyjson.com/image/i/products/23/thumbnail.jpg"
      ]
    },
    {
      "id": 24,
      "title": "cereals muesli fruit nuts",
      "description": "original fauji cereal muesli 250gm box pack original fauji cereals muesli fruit nuts flakes breakfast cereal break fast faujicereals cerels cerel foji fouji",
      "price": 46,
      "rating": { "rate": 4.94, "count": 11378 },
      "stock": 113,
      "category": "groceries",
      "images": [
        "https://dummyjson.com/image/i/products/24/1.jpg",
        "https://dummyjson.com/image/i/products/24/2.jpg",
        "https://dummyjson.com/image/i/products/24/3.jpg",
        "https://dummyjson.com/image/i/products/24/4.jpg",
        "https://dummyjson.com/image/i/products/24/thumbnail.jpg"
      ],
      "lastModified": "2022-10-17T14:31:36.884Z"
    },
    {
      "id": 25,
      "title": "Gulab Powder 50 Gram",
      "description": "Dry Rose Flower Powder Gulab Powder 50 Gram â€¢ Treats Wounds",
      "price": 70,
      "rating": { "rate": 4.87, "count": 568 },
      "stock": 47,
      "category": "groceries",
      "images": [
        "https://dummyjson.com/image/i/products/25/1.png",
        "https://dummyjson.com/image/i/products/25/2.jpg",
        "https://dummyjson.com/image/i/products/25/3.png",
        "https://dummyjson.com/image/i/products/25/4.jpg",
        "https://dummyjson.com/image/i/products/25/thumbnail.jpg"
      ]
    }
  ]
  