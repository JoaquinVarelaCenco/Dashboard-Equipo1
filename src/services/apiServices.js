const url = "http://localhost:5000/api/";
const getProducts = () => {
  return fetch(`${url}product`).then((res) => res.json())
};

const getProductById = (id) => {
  return fetch(`${url}product/${id}`).then((res) => res.json());
};

const deleteProduct = (productId) => {
  return fetch(`${url}product?id=${productId}`, {
    method: "DELETE",
  }).then((res) => res.json());
};

const updateProduct = (body) => {
  return fetch(`${url}product`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((res) => res.json());
};

const newProduct = (body) => {
  return fetch(`${url}product`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((res) => res.json());
};

module.exports = {
  getProducts,
  getProductById,
  deleteProduct,
  updateProduct,
  newProduct,
};
