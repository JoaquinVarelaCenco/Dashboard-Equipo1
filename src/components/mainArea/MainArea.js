import "./MainArea.css";
import Header from "../header/Header"
import AppRouter from "../../routes/AppRouter"
import { useContext } from "react";
import { ThemeContext } from "../../context/ThemeContext";

const MainArea = () => {
  const { theme } = useContext(ThemeContext);

  return (
    <div className={`mainArea ${theme}`}>
      <Header/>
      <div className="mainArea__content">
      <AppRouter>
      </AppRouter>
      </div>
      </div> 
      
  );
};

export default MainArea;
