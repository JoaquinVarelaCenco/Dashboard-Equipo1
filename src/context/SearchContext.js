import { useState, createContext, useEffect } from "react";
import { getProducts } from "../services/apiServices";

export const SearchContext = createContext();

export const SearchProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [productsExist, setProductsExist] = useState(true);
  const [helpProducts, setHelpProducts] = useState([]);
  const [searchTermValue, setSearchTermValue] = useState("");
  const [orderBy, setOrderBy] = useState("Mas relevantes");

  const getAllProducts = () => {
    getProducts()
      .then((data) => {
        setProductsExist(true);
        orderProductsPrueba(data);
        setProducts(data);
        setHelpProducts(data);
      })
      .catch(() => {
        setProductsExist(false);
      });
  };

  const filterProducts = (searchTerm) => {
    const newProducts = helpProducts.filter((val) => {
      if (
        searchTerm === "" ||
        val.title.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      } else if (
        val.description.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      }
    });
    return newProducts;
  };

  const handleSearch = (e) => {
    const newFilter = filterProducts(e.target.value);
    setSearchTermValue(e.target.value);
    setProducts(newFilter);
  };

  const orderByHighPrice = (arrayData) => {
    arrayData.sort((a, b) =>
      a.price < b.price ? 1 : a.price > b.price ? -1 : 0
    );
  };

  const orderByLowPrice = (arrayData) => {
    arrayData.sort((a, b) =>
      a.price > b.price ? 1 : a.price < b.price ? -1 : 0
    );
  };

  const orderByMostRelevants = (arrayData) => {
    arrayData.sort((a, b) =>
      a.rating.count < b.rating.count
        ? 1
        : a.rating.count > b.rating.count
        ? -1
        : 0
    );
  };

  const orderByAlphabet = (arrayData) => {
    arrayData.sort((a, b) =>
      a.title > b.title ? 1 : a.title < b.title ? -1 : 0
    );
  };

  const orderProducts = (e) => {
    switch (e.target.value) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        orderByHighPrice(products);
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        orderByLowPrice(products);
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        orderByMostRelevants(products);
        break;
      case "A-Z":
        setOrderBy("A-Z");
        orderByAlphabet(products);
        break;
    }
  };

  const orderProductsPrueba = (data) => {
    
    switch (orderBy) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        orderByHighPrice(data);
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        orderByLowPrice(data);
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        orderByMostRelevants(data);
        break;
      case "A-Z":
        setOrderBy("A-Z");
        orderByAlphabet(data);
        break;
    }
  };



  useEffect(() => {
    getAllProducts();
  }, []);

  return (
    <SearchContext.Provider
      value={{
        handleSearch,
        products,
        productsExist,
        getAllProducts,
        helpProducts,
        searchTermValue,
        orderBy,
        orderProducts
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};
