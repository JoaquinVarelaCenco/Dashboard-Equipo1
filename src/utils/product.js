import { deleteProduct } from "../services/apiServices";


export function deleteProductFunction(id) {
    deleteProduct(id).then((res) => {
    });
}

export const sortByCount = (products) =>{
    const sorted = products.sort((a,b) =>{
    if(a.rating.count < b.rating.count){
        return 1;
    }else  if(a.rating.count > b.rating.count){
        return -1;
    }else{
        return 0;
    }
    });
    return sorted;
}

export const sortByStock = (products) =>{
    const sorted = products.sort((a,b) =>{
    if(a.stock < b.stock){
        return 1;
    }else  if(a.stock > b.stock){
        return -1;
    }else{
        return 0;
    }
    });
    return sorted;
}