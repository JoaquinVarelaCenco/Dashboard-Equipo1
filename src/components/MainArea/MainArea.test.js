import { render, act, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import { HeaderProvider } from "../../context/HeaderContext";
import { SearchProvider } from "../../context/SearchContext";
import { SideBarProvider } from "../../context/SideBarContext";
import productsData from "../../mockData/productsData";
import { getProducts } from "../../services/apiServices";
import { orderByMostRelevants } from "../../utils/product";
import MainArea from "./MainArea";

jest.mock("../../services/apiServices");

const filterProducts = (searchTerm, arrayProd) => {
    const newProducts = arrayProd.filter((val) => {
      if (
        searchTerm === "" ||
        val.title.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      } else if (
        val.description.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      }
    });
    return newProducts;
  };

describe("Test formulario de productos", () => {
  let component;

  beforeEach(async () => {
    getProducts.mockResolvedValue([...productsData]);

    await act(async () => {
      component = await render(
        <MemoryRouter initialEntries={["/products"]}>
          <HeaderProvider>
            <SideBarProvider>
              <SearchProvider>
                <MainArea />
              </SearchProvider>
            </SideBarProvider>
          </HeaderProvider>
        </MemoryRouter>
      );
    });
  });

  it("renderiza el componente", () => {
    let { container } = component;
    expect(container).toMatchSnapshot();
  });

  it("Se debe poder escribir en el input de busqueda y su valor debe ser el ingresado", ()=>{
    let testWord = "test";
    let inputSearch = screen.getByRole('textbox')
    userEvent.type(inputSearch, testWord);
    expect(inputSearch).toHaveValue(testWord);
})

it("Se deben renderizar todos los productos cuando el input de busqueda no se ingresaron caracteres", ()=>{
    const arrayProd = screen.getAllByTestId('id-product');
    let arrayProdPivot = arrayProd.map(a => a.innerHTML);
    let returnProductsData = orderByMostRelevants(productsData);
    returnProductsData = returnProductsData.map(b => `#${b.id}`);
    expect(returnProductsData).toEqual(arrayProdPivot);
})

it("Se deben renderizar las coincidencias de los productos con lo escrito en el input de busqueda", ()=>{
    let testWord = "iphone";
    let inputSearch = screen.getByRole('textbox')
    userEvent.type(inputSearch, testWord);
    expect(inputSearch.value).toBe(testWord);
    const arrayProd = screen.getAllByTestId('id-product');
    let arrayProdPivot = arrayProd.map(a => a.innerHTML);
    let orderProductsData = filterProducts(testWord, productsData).map(b => `#${b.id}`);
    expect(orderProductsData).toEqual(arrayProdPivot);
})

});
