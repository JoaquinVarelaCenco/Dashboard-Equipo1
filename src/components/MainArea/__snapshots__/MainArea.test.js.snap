// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test formulario de productos renderiza el componente 1`] = `
<div>
  <div
    class="mainArea undefined"
  >
    <div
      class="header undefined"
    >
      <div
        class="headerGeneric displayFlex"
      >
        <button
          class="header__hamburguer-menu-btn noHoverEffect"
        >
          <img
            alt="MenÃº hamburguesa"
            class="header__hamburguer-menu-img"
            src="menu.svg"
          />
        </button>
        <div
          class="title"
        >
          <a
            href="/products"
          >
            <h2>
              Productos
            </h2>
          </a>
        </div>
      </div>
      <div
        class="headerProducts"
      >
        <div
          class="headerProducts__search-container   closeSearchBarAnimation "
        >
          <button
            class="headerProducts-search-container__btnClose headerBtn hideComponent noHoverEffect"
          >
            X
          </button>
          <input
            class="header__search  "
            placeholder=""
            type="text"
          />
          <button
            class="headerProducts-search-container__btnSearch headerBtn noHoverEffect"
          >
            <img
              alt="Lupa de busqueda"
              class="headerProducts-search-container__btnImage"
              src="magnify.svg"
            />
          </button>
        </div>
        <div
          class="headerProducts-ContainerAdd"
        >
          <a
            href="/products/new"
          >
            <button
              class="headerProducts__btnAdd"
            >
              Agregar Producto
            </button>
          </a>
        </div>
        <div
          class="headerProducts-ContainerAdd"
        >
          <a
            href="/products/new"
            style="display: block;"
          >
            <button
              class="headerProducts__btnAddAlternative noHoverEffect"
            >
              +
            </button>
          </a>
        </div>
      </div>
    </div>
    <div
      class="mainArea__content"
    >
      <div
        class="productList__container"
      >
        <div
          class="productList__container__sort"
        >
          <div>
            <i>
              Ordenar por
            </i>
            <select
              class="select-control-products"
              name="category"
            >
              <option
                value="Mayor precio"
              >
                Mayor precio
              </option>
              <option
                value="Menor precio"
              >
                Menor precio
              </option>
              <option
                value="Mas relevantes"
              >
                Mas relevantes
              </option>
              <option
                value="A-Z"
              >
                A-Z
              </option>
            </select>
          </div>
          <i>
            5
             resultados
          </i>
        </div>
        <a
          class="product-card-link"
          href="/products/24"
        >
          <article
            class="main-product__related-product-card"
          >
            <div
              class="main-product__imageAndDesc"
            >
              <div
                class="section-article__image"
              >
                <img
                  alt="Imagen de producto"
                  class="imagenPrincipal hideImage"
                  src="https://dummyjson.com/image/i/products/24/1.jpg"
                />
                <div
                  class="spinnerWrapper showSpinner"
                >
                  <div
                    class="spinnerContainer"
                  >
                    <div
                      class="lds-ripple"
                    >
                      <div />
                      <div />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="section-article__desc"
              >
                <div
                  class="desc"
                >
                  <p>
                    cereals muesli fruit nuts
                  </p>
                </div>
                <div
                  class="section-article__price"
                >
                  <p
                    class="valorNumerico"
                    data-testid="id-product"
                  >
                    #
                    24
                  </p>
                </div>
              </div>
            </div>
            <div
              class="section-article__goToProduct"
            >
              <img
                alt=""
                src="chevron-right-test.svg"
              />
            </div>
          </article>
        </a>
        <a
          class="product-card-link"
          href="/products/23"
        >
          <article
            class="main-product__related-product-card"
          >
            <div
              class="main-product__imageAndDesc"
            >
              <div
                class="section-article__image"
              >
                <img
                  alt="Imagen de producto"
                  class="imagenPrincipal hideImage"
                  src="https://dummyjson.com/image/i/products/23/1.jpg"
                />
                <div
                  class="spinnerWrapper showSpinner"
                >
                  <div
                    class="spinnerContainer"
                  >
                    <div
                      class="lds-ripple"
                    >
                      <div />
                      <div />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="section-article__desc"
              >
                <div
                  class="desc"
                >
                  <p>
                    Orange Essence Food Flavou
                  </p>
                </div>
                <div
                  class="section-article__price"
                >
                  <p
                    class="valorNumerico"
                    data-testid="id-product"
                  >
                    #
                    23
                  </p>
                </div>
              </div>
            </div>
            <div
              class="section-article__goToProduct"
            >
              <img
                alt=""
                src="chevron-right-test.svg"
              />
            </div>
          </article>
        </a>
        <a
          class="product-card-link"
          href="/products/3"
        >
          <article
            class="main-product__related-product-card"
          >
            <div
              class="main-product__imageAndDesc"
            >
              <div
                class="section-article__image"
              >
                <img
                  alt="Imagen de producto"
                  class="imagenPrincipal hideImage"
                  src="no-image.jpeg"
                />
                <div
                  class="spinnerWrapper showSpinner"
                >
                  <div
                    class="spinnerContainer"
                  >
                    <div
                      class="lds-ripple"
                    >
                      <div />
                      <div />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="section-article__desc"
              >
                <div
                  class="desc"
                >
                  <p>
                    Samsung Universe 9
                  </p>
                </div>
                <div
                  class="section-article__price"
                >
                  <p
                    class="valorNumerico"
                    data-testid="id-product"
                  >
                    #
                    3
                  </p>
                </div>
              </div>
            </div>
            <div
              class="section-article__goToProduct"
            >
              <img
                alt=""
                src="chevron-right-test.svg"
              />
            </div>
          </article>
        </a>
        <a
          class="product-card-link"
          href="/products/25"
        >
          <article
            class="main-product__related-product-card"
          >
            <div
              class="main-product__imageAndDesc"
            >
              <div
                class="section-article__image"
              >
                <img
                  alt="Imagen de producto"
                  class="imagenPrincipal hideImage"
                  src="https://dummyjson.com/image/i/products/25/1.png"
                />
                <div
                  class="spinnerWrapper showSpinner"
                >
                  <div
                    class="spinnerContainer"
                  >
                    <div
                      class="lds-ripple"
                    >
                      <div />
                      <div />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="section-article__desc"
              >
                <div
                  class="desc"
                >
                  <p>
                    Gulab Powder 50 Gram
                  </p>
                </div>
                <div
                  class="section-article__price"
                >
                  <p
                    class="valorNumerico"
                    data-testid="id-product"
                  >
                    #
                    25
                  </p>
                </div>
              </div>
            </div>
            <div
              class="section-article__goToProduct"
            >
              <img
                alt=""
                src="chevron-right-test.svg"
              />
            </div>
          </article>
        </a>
        <a
          class="product-card-link"
          href="/products/2"
        >
          <article
            class="main-product__related-product-card"
          >
            <div
              class="main-product__imageAndDesc"
            >
              <div
                class="section-article__image"
              >
                <img
                  alt="Imagen de producto"
                  class="imagenPrincipal hideImage"
                  src="no-image.jpeg"
                />
                <div
                  class="spinnerWrapper showSpinner"
                >
                  <div
                    class="spinnerContainer"
                  >
                    <div
                      class="lds-ripple"
                    >
                      <div />
                      <div />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="section-article__desc"
              >
                <div
                  class="desc"
                >
                  <p>
                    iPhone X
                  </p>
                </div>
                <div
                  class="section-article__price"
                >
                  <p
                    class="valorNumerico"
                    data-testid="id-product"
                  >
                    #
                    2
                  </p>
                </div>
              </div>
            </div>
            <div
              class="section-article__goToProduct"
            >
              <img
                alt=""
                src="chevron-right-test.svg"
              />
            </div>
          </article>
        </a>
      </div>
    </div>
  </div>
</div>
`;
